### PR TITLE
feat(router): add and remove `initialNavigation` options

### PR DESCRIPTION
**Pull request by CaerusKaru**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: https://github.com/angular/angular/issues/32887

## What is the new behavior?

Added two new options, removed four old options, as described in above issue.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

Added migration for removed options

### CODE REVIEW

**Comment by CaerusKaru** [comment]
cc @vikerman @IgorMinar @devversion

File: packages/core/schematics/migrations.json [hunk 1]
```diff
@@ -39,6 +39,11 @@
       "version": "9-beta",
       "description": "Removes the `static` flag from dynamic queries.",
       "factory": "./migrations/dynamic-queries/index"
+    },
+    "migration-v9-initial-navigation": {
+      "version": "9.0.0-beta",
+      "description": "Updates the `initialNavigation` property for `RouterModule.forRoot`.",
+      "factory": "./migrations/initial-navigation/index"
```
Comment (devversion) [inline packages/core/schematics/migrations.json]: Should the migration name give more context? e.g. `router-initial-navigation-options`

File: packages/core/schematics/migrations/initial-navigation/util.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  const importSpecifier = findImportSpecifier(routerImport.elements, routerModule);
+  let routerModuleNode: ts.CallExpression|undefined = undefined;
+
+  if (importSpecifier) {
+    ts.forEachChild(sourceFile, function visitNode(node: ts.Node) {
+      if (ts.isCallExpression(node) && ts.isPropertyAccessExpression(node.expression) &&
+          ts.isIdentifier(node.expression.expression) &&
+          node.expression.expression.text === importSpecifier.name.text &&
+          node.expression.name.text === 'forRoot') {
+        routerModuleNode = node;
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/util.ts]: I'm trying to understand why we search instances of the `RouterModule` based on the import. An easier solution where we could just leverage our existing utilities (e.g. `getImportOfIdentifier`) would be more robust as it could also handle aliased imports (and eventually namespaced as well).

Basically, most of the logic remains the same, but we just walk through the source file _once_ and look for `ts.Identifier`'s which we then run against `getImportOfIdentifier` to check if the property name is `RouterModule` and the module name is `@angular/router`. 

I realize this is used in one of the other migrations as well, but that is not robust either. Just checking identifier names is not reliable.

File: packages/core/schematics/migrations/initial-navigation/util.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+      ts.forEachChild(node, visitNode);
+    });
+  }
+
+  return routerModuleNode;
+}
+
+export function updateInitialNavigation(node: ts.CallExpression): ts.CallExpression {
+  const existingOptions = node.arguments[1] as ts.ObjectLiteralExpression | undefined;
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/util.ts]: It would be probably better to just check if `node.arguments[1]` is actually a `ts.ObjectLiteralExpression`. Some applications pass in an identifier and that would cause a migration runtime exception down the line.

e.g.

```ts
export const ROUTER_CONFIG = ....

..

RouterModule.forRoot(.., ROUTER_CONFIG)
```

File: packages/core/schematics/migrations/initial-navigation/util.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  const initialNavigation = 'initialNavigation';
+
+  if (existingOptions) {
+    const properties = existingOptions.properties.map(exp => {
+      if (ts.isPropertyAssignment(exp) && ts.isIdentifier(exp.name) &&
+          exp.name.text === initialNavigation) {
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/util.ts]: To handle all statically analyzable cases of the property name, you could use `getPropertyNameText(exp.name) === initialNavigation)` from the schematic utils.

File: packages/core/schematics/migrations/initial-navigation/util.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        return ts.updatePropertyAssignment(
+            exp, ts.createStringLiteral(initialNavigation),
+            ts.createStringLiteral(updatedNavigation));
+      }
+
+      return exp;
+    });
+
+    const args = [node.arguments[0], ts.updateObjectLiteral(existingOptions, properties)];
+    return ts.createCall(node.expression, node.typeArguments, args);
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/util.ts]: I'd just return the updated object literal for the options and print that to the source file. Printing the whole decorator is unnecessary and causes unnecessary formatting of the routes literal (and in this case even causes loss of comments)

**Review by vikerman — APPROVED** [review]
LGTM for the the actual router changes.

Are there any other doc changes required to mention enabledNonBlocking as the default?

**Comment by Denis-Evseev** [comment]
Any ideas when it is gonna be released?

**Comment by Splaktar** [comment]
@Denis-Evseev I'm not 100% sure, but I think that this got bumped from v9 to v10.

**Comment by CaerusKaru** [comment]
@Splaktar @Denis-Evseev This PR will not be included as part of v9, since it was decided that there wasn't enough time to announce, test, and integrate this change. It will be part of v10. The changes in NgUniversal are unrelated to this PR.

**Comment by SchnWalter** [comment]
> LGTM for the the actual router changes.
> 
> Are there any other doc changes required to mention enabledNonBlocking as the default?

I think that these phrases need to be more clear:
* > initial navigation starts before the root component is created
* > initial navigation starts after the root component has been created

For example, in Cordova, you need to wait for the ready event before you can start using the available APIs and the `APP_INITIALIZER` is the appropriate place to add a "delay". But if you do this and your guards depend on the APIs, you'll soon realize that you have a race condition where the `initialNavigation` from `RouterInitializer` sometimes starts to consume the guards before your custom `APP_INITIALIZER` finishes. And this is OK if you think about the various `APP_INITIALIZER` that run in paralel, but it's not obvious from that phrase that this is what's happening.

File: packages/core/schematics/migrations.json [hunk 6]
```diff
@@ -39,6 +39,11 @@
       "version": "9-beta",
       "description": "Removes the `static` flag from dynamic queries.",
       "factory": "./migrations/dynamic-queries/index"
+    },
+    "migration-v9-initial-navigation": {
```
Comment (atscott) [inline packages/core/schematics/migrations.json:43]: Nit: please also update the commit message to be a little bit more descriptive

File: packages/router/src/router_module.ts [hunk 7]
```diff
@@ -229,28 +229,22 @@ export function provideRoutes(routes: Routes): any {
...
- * The bootstrap is not blocked until the initial navigation is complete. @deprecated
- * * 'legacy_disabled'- The initial navigation is not performed. The location listener is set up
- * after the root component gets created. @deprecated since v4
- * * `true` - same as 'legacy_enabled'. @deprecated since v4
- * * `false` - same as 'legacy_disabled'. @deprecated since v4
- *
- * The 'legacy_enabled' and 'legacy_disabled' should not be used for new applications.
+ * * 'enabledNonBlocking' - (Default, for compatibility.) The initial navigation starts after the root component has been created.
+ * The bootstrap is not blocked until the initial navigation is complete.
+ * * 'enabled' - Deprecated, same as `enabledBlocking`, will be removed in v10
```
Comment (atscott) [inline packages/router/src/router_module.ts:241]: Update comment to "Will be removed in v11"? And does this need `@deprecated`

**Review by atscott — CHANGES_REQUESTED** [review]
Added a few comments. Can you address the comments from @devversion as well?

File: packages/router/src/router_module.ts (hunk 7 above)
Comment (CaerusKaru) [reply packages/router/src/router_module.ts:242]: I think if we mark any part of the comment as deprecated, it gets applied to the entire block.

File: packages/router/src/router_module.ts [hunk 8]
```diff
@@ -532,9 +522,8 @@ export class RouterInitializer {
           }
         };
         router.initialNavigation();
-
       } else {
-        throw new Error(`Invalid initialNavigation options: '${opts.initialNavigation}'`);
+        resolve(true);
```
[the remark this answers is not in the document]
Comment (CaerusKaru) [reply packages/router/src/router_module.ts:528]: This case covers `enabledNonBlocking` and `undefined` (since both are considered default).

File: packages/router/src/router_module.ts [hunk 9]
```diff
@@ -552,10 +541,8 @@ export class RouterInitializer {
       return;
     }
 
-    if (this.isLegacyEnabled(opts)) {
+    if (opts.initialNavigation === 'enabledNonBlocking') {
       router.initialNavigation();
-    } else if (this.isLegacyDisabled(opts)) {
-      router.setUpLocationChangeListener();
     }
```
[the remark this answers is not in the document]
Comment (CaerusKaru) [reply packages/router/src/router_module.ts:548]: The other case is `undefined`, I think it's clearer to have a drop case, like a `default` in a `switch`

File: packages/core/schematics/migrations/initial-navigation/collector.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+import {getImportOfIdentifier} from '../../utils/typescript/imports';
+
+
+/**
+ * Visitor that walks through specified TypeScript nodes and collects all
+ * found RouterModule instances that have `forRoot` call expressions
+ */
+export class Collector {
+  public resolvedModules: ts.CallExpression[] = [];
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/collector.ts:18]: Can we rename the class to something more specific? Also change the variable to something like `routerModuleSetupCalls`? Personally I'd expect `resolvedModules` to refer to actual identifiers of `RouterModule` (which is obviously not the case 😄 )

File: packages/core/schematics/migrations/initial-navigation/collector.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    ts.forEachChild(node, n => this.visitNode(n));
+  }
+}
+
+function isRouterModuleCall(typeChecker: ts.TypeChecker, node: ts.Node): node is ts.CallExpression {
+  const moduleName = '@angular/router';
+  const routerModule = 'RouterModule';
+  const functionCall = 'forRoot';
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/collector.ts:34]: Nit: Move to constants at the top-level of the file?

File: packages/core/schematics/migrations/initial-navigation/collector.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  if (ts.isCallExpression(node) && ts.isPropertyAccessExpression(node.expression) &&
+      ts.isIdentifier(node.expression.expression)) {
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/collector.ts:36]: This does not handle namespace imports. e.g. `router.RouterModule.forRoot()`. I have plans to expand the import utilities to handle this (as we do in the components repo). Maybe leave a small TODO here? Alternatively, we should evaluate if this is something we want to handle for this migration?

File: packages/core/schematics/migrations/initial-navigation/collector.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    const module = node.expression.expression;
+    const call = node.expression.name;
+    const imp = getImportOfIdentifier(typeChecker, module);
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/collector.ts:39]: Optional nit for more obvious/correct names:

```suggestion
    const containerExpr = node.expression.expression;
    const functionName = node.expression.name.text;
    const importData = getImportOfIdentifier(typeChecker, module);
```

File: packages/core/schematics/migrations/initial-navigation/index.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    tree: Tree, tsconfigPath: string, basePath: string): string[] {
+  const {program} = createMigrationProgram(tree, tsconfigPath, basePath);
+  const failures: string[] = [];
+  const typeChecker = program.getTypeChecker();
+  const definitionCollector = new Collector(typeChecker);
+  const sourceFiles = program.getSourceFiles().filter(
+      f => !f.isDeclarationFile && !program.isSourceFileFromExternalLibrary(f));
+
+  // Analyze source files by detecting all modules, directives and components.
+  sourceFiles.forEach(sourceFile => definitionCollector.visitNode(sourceFile));
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/index.ts:53]: The comment above is no longer correct (mostly a leftover from other schematics). Can we update and also not call it definition collector?

File: packages/core/schematics/test/initial_navigation_migration_spec.ts [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    await runMigration();
+    expect(tree.readContent('/index.ts')).toContain(`initialNavigation: "enabledNonBlocking"`);
+  });
+
+  function writeFile(filePath: string, contents: string) {
+    host.sync.write(normalize(filePath), virtualFs.stringToFileBuffer(contents));
+  }
+
+  function runMigration() {
+    runner.runSchematicAsync('migration-v10-router-initial-navigation-options', {}, tree)
```
Comment (devversion) [inline packages/core/schematics/test/initial_navigation_migration_spec.ts:126]: Is this missing an await or return?

File: packages/core/schematics/migrations/initial-navigation/transform.ts [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+  private optionsEvaluator: OptionsEvaluator = new OptionsEvaluator(
+      new TypeScriptReflectionHost(this.typeChecker), this.typeChecker,
+      /* dependencyTracker */ null);
+
+  constructor(
+      private typeChecker: ts.TypeChecker,
+      private getUpdateRecorder: (sf: ts.SourceFile) => UpdateRecorder) {}
+
+  /** Migrate the RouterModule.forRoot call expression objects. */
+  migrateModules(modules: ts.CallExpression[]): AnalysisFailure[] {
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/transform.ts:35]:
```suggestion
  migrateRouterSetupCalls(modules: ts.CallExpression[]): AnalysisFailure[] {
```

File: packages/core/schematics/migrations/initial-navigation/transform.ts [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+      if (typeof mapValue === 'string' || typeof mapValue === 'boolean') {
+        const props = node.properties.filter(
+            p => ts.isPropertyAssignment(p) && ts.isIdentifier(p.name) &&
+                p.name.text !== initialNavigation);
+        this._updateNode(
+            node,
+            ts.updateObjectLiteral(
+                node,
+                props.concat(ts.createPropertyAssignment(
+                    initialNavigation, ts.createStringLiteral(updateText(mapValue))))));
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/transform.ts:72]: I don't think we _always_ want to update the object literal here? We could unnecessarily throw of formatting or cause the devkit to write file changes. We can only update if the value needs to be updated, right?

File: packages/core/schematics/migrations/initial-navigation/transform.ts [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+        return [];
+      } else if (mapValue instanceof DynamicValue) {
+        addTodoToNode(mapValue.node, TODO_COMMENT);
+        this._addComment(mapValue.node);
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/transform.ts:76]: Aren't we adding two comments here?

File: packages/core/schematics/migrations/initial-navigation/transform.ts [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+    recorder.updateObjectLiteralExpression(node, newText);
+  }
+}
+
+/**
+ * Updates the deprecated initialNavigation options to their v10 equivalents
+ * (or as close as we can get).
+ * @param text the old text to update
+ */
+function updateText(text: string|boolean) {
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/transform.ts:105]: Maybe give this a more specific name? e.g. `getUpdatedInitialNavigationValue`

File: packages/core/schematics/migrations/initial-navigation/transform.ts [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+
+  /**
+   * Migrate a RouterModule.forRoot call expression to use the new options format for
+   * initialNavigation
+   */
+  migrateModule(module: ts.CallExpression): AnalysisFailure[] {
+    const options = module.arguments[1];
+    if (options) {
+      const {resolvedValue, node} = this.optionsEvaluator.evaluate(options);
+      return this._getResolvedObjectLiteral(resolvedValue, node);
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/transform.ts:48]:
```ts
      // Updates the router setup options if needed. Returns a list of analysis
      // failures if migration could not be performed.
      return this._updateRouterSetupOptions(resolvedValue, node);
```

File: packages/core/schematics/test/initial_navigation_migration_spec.ts [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+          ]
+        })
+        export class AppModule {
+        }
+      `);
+
+    await runMigration();
+    expect(tree.readContent('/index.ts')).toContain(`initialNavigation: "enabledNonBlocking"`);
+  });
+
```
Comment (devversion) [inline packages/core/schematics/test/initial_navigation_migration_spec.ts:120]: Can we have tests for the TODO logic (which seems to be broken right now). Also a test for object spread would be good (as that logic seems currently broken)

File: packages/core/schematics/migrations/initial-navigation/options_evaluator.ts [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+ * callee scope). This can be avoided by simply extending the static interpreter and intercepting
+ * the "visitObjectLiteralExpression" method.
+ */
+export class OptionsEvaluator extends StaticInterpreter {
+  private _optionsNode?: ts.ObjectLiteralExpression;
+
+  visitObjectLiteralExpression(node: ts.ObjectLiteralExpression, context: any) {
+    const resolvedValue =
+        super.visitObjectLiteralExpression(node, {...context, insideOptionsDef: true});
+    this._optionsNode = node;
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/options_evaluator.ts:29]: Looks like this is not handling `insideOptionsDef` properly. i.e. it will not work as expected if the interpreter visits multiple object literals (due to spreads for example).

What is the goal here? The `initialNavigation` option could be defined anywhere and it will be difficult to the detect the actual property node. Maybe it's best to just have the following logic:

1. Evaluate options object.
2. Check if `initialNavigation` is set. If not, abort.
3. Check if the initial navigation value needs an update. If not, abort.
4. Access the first found object literal and migrate if it has the `initialNavigation` prop. If not, then it will difficult to migrate and we could just propose the update as a comment/failure?

We could definitely improve (4) with some advanced traversing of the TS nodes to find the right property that controls `initialNavigation`. e.g. using the last object literal that has `initialNavigation` (as the last one usually overrides previous in a spread; might need some double-checking for edge cases).

File: packages/core/schematics/migrations/initial-navigation/transform.ts [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+    return [];
+  }
+
+  /**
+   * Visits a given object literal expression to update the initialNavigation property. If the
+   * expression cannot be resolved, add a TODO to alert the user.
+   */
+  private _getResolvedObjectLiteral(value: ResolvedValue, node?: ts.ObjectLiteralExpression):
+      AnalysisFailure[] {
+    const initialNavigation = 'initialNavigation';
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/transform.ts:60]: Optional nit: Move to top-level constant.

File: packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Replacement, RuleFailure} from 'tslint';
+import * as ts from 'typescript';
+
+import {UpdateRecorder} from '../update_recorder';
+
+export class TslintUpdateRecorder implements UpdateRecorder {
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: Do we know whether a google3 tslint driver is needed? I could imagine this being a somewhat advanced, but really fast find/replace internally.

**Review by atscott — APPROVED** [review]
LGTM for the router API changes. Deferring to @devversion for review of migrations.

Reviewed for: fw-router

File: packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts (hunk 24 above)
Comment (CaerusKaru) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: google3 tests failed, and we're under a time crunch, so I'm hoping this will make it easier.

File: packages/core/schematics/migrations.json [hunk 25]
```diff
@@ -44,6 +44,11 @@
       "version": "9.0.0-beta",
       "description": "ModuleWithProviders migration. In Angular 9, the ModuleWithProviders type without a generic has been deprecated. This migration adds the generic where it is missing. Read more about this here: https://v9.angular.io/guide/migration-module-with-providers",
       "factory": "./migrations/module-with-providers/index"
+    },
+    "migration-v10-router-initial-navigation-options": {
+      "version": "10.0.0-beta",
+      "description": "Updates the `initialNavigation` property for `RouterModule.forRoot`.",
```
[the remark this answers is not in the document]
Comment (CaerusKaru) [reply packages/core/schematics/migrations.json:50]: Not for this one, it's a holdover from v9 when it was agreed upon, and there wasn't a plan then either.

File: packages/core/schematics/migrations/initial-navigation/transform.ts (hunk 18 above)
Comment (CaerusKaru) [reply packages/core/schematics/migrations/initial-navigation/transform.ts:83]: No, `addTodoToNode` modifies the `ts.Node` itself, `_addComment` actually commits that change.

File: packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts (hunk 24 above)
Comment (devversion) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: Works for me. Do we have any data? Caretaker should know how many instances there are when the presubmit runs. cc. @atscott is there already one?

File: packages/core/schematics/migrations/initial-navigation/options_evaluator.ts (hunk 22 above)
Comment (devversion) [reply packages/core/schematics/migrations/initial-navigation/options_evaluator.ts]: Based on discussion with Adam, it's probably easier to just look for all object literals with `initialNavigation`. If they use a deprecated value, we could migrate them, or add a comment.

I think in some cases this is not statically analyzable. We could still use the partial evaluator for those, but I think if a `DynamicValue` is resolved, then it doesn't help us as it's still unclear whether this refers to an property assignment for the router `initialNavigation` option.

File: packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts (hunk 24 above)
Comment (atscott) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: Already a what? Instances there are of what? Sorry, I'm not familiar with what's being discussed here.

File: packages/core/schematics/migrations/initial-navigation/transform.ts (hunk 18 above)
Comment (devversion) [reply packages/core/schematics/migrations/initial-navigation/transform.ts:83]: That seems really confusing. Can we move `addTodoToNode` into `_addComment`. And then in the update recorders we rename the method to `updateNode(existingNode: ts.Node, newText: string)`. For tslint the description can be just: `Node needs to be updated to: ${newText}` (like we did in other migrations)

File: packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts (hunk 24 above)
Comment (CaerusKaru) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: I think Paul's asking if there are any documented failures from a prior presubmit run, that way we can judge how much this is needed

Comment (devversion) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: I'm asking whether there is already a presubmit for this. We are trying to figure out how many instances of `initialNavigation: <..>` are in g3 that rely on the removed values.

I'm trying to figure out whether we need to run the migration inside google3. usually these take up a lot of time to run in g3 (based on what I heard), and a simple find/replace is easier from what I heard.

Comment (atscott) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: Ah, got it. Any old runs would be expired by now. Started new [presubmit](http://test/OCL:300562098:BASE:308873304:1588101767110:2de63cf4).
Also started global presubmit, as that would be needed as well (accessible via the same link above).

Comment (atscott) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: I also did a codesearch for "initialnavigation: " and only got 34 results.

Comment (devversion) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: Thanks Andrew! 34 results aren't a lot I guess. Depending on how the presubmit turns out, it seems like the g3 is not needed.

Comment (atscott) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: Yea, I don't think it would be too difficult to update the values by hand. The only difficulty would be getting all the necessary approvals for submitting the g3 sync. 

I'm not sure what the expectation is for the presubmit. Wouldn't we just have everything that depends on the `initialNavigation` value simply fail? Unless we update those values to match the new ones, I don't think we'll have meaningful results outside of verifying that the default (`undefined`) has no change in behavior. I feel like I'm missing some context here...

File: packages/core/schematics/migrations/initial-navigation/options_evaluator.ts [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+    if (resolvedValue instanceof Map && resolvedValue.get('initialNavigation') !== undefined) {
+      this._optionsNodes.push({node, resolvedValue});
+    }
+
+    return resolvedValue;
+  }
+
+  /**
+   * Evaluates the given expression and returns its statically resolved value
+   * and a list of object literals which define Angular providers.
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/options_evaluator.ts:43]: This commit needs to be updated. It's still based on the missing-injectable evaluator.

File: packages/core/schematics/test/google3/initial_navigation_rule_spec.ts [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+  it('should migrate nested objects opposite order', () => {
+    writeFile('/index.ts', `
+        import { NgModule } from '@angular/core';
+        import { RouterModule } from '@angular/router';
+
+        const options = {initialNavigation: 'legacy_enabled'};
+
+        @NgModule({
+          imports: [
+            RouterModule.forRoot([], {...options, initialNavigation: 'legacy_disabled'}),
```
Comment (devversion) [inline packages/core/schematics/test/google3/initial_navigation_rule_spec.ts:172]: I think with the current logic, cases like the following are not migrated properly:

```ts
const defaultOptions: ExtraOptions = {initialNavigation: 'legacy_disabled'};
RouterModule.forRoot([], {...defaultOptions, initialNavigation: 'enabledNonBlocking'})
```

The resolved value will hide `legacy_disabled` and the migration will be skipped. Though, ideally the migration would update the `defaultOptions` literal too.

**Review by devversion — COMMENTED** [review]
The router changes LGTM too. For the migration, the code looks good too, but it looks like the current migration strategy does not capture a few potentially common cases.

File: packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts (hunk 24 above)
Comment (atscott) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: [global presubmit](https://test.corp.google.com/ui#id=OCL:308924214:BASE:309243397:1588265081380:797dc1f8) with modifications made by hand to update `initialNavigation` in g3. Everything builds, but there are still 109 failing tests that need investigation.

File: packages/router/src/router_module.ts [hunk 28]
```diff
@@ -229,28 +229,23 @@ export function provideRoutes(routes: Routes): any {
...
- * * 'legacy_disabled'- The initial navigation is not performed. The location listener is set up
- * after the root component gets created. @deprecated since v4
- * * `true` - same as 'legacy_enabled'. @deprecated since v4
- * * `false` - same as 'legacy_disabled'. @deprecated since v4
- *
- * The 'legacy_enabled' and 'legacy_disabled' should not be used for new applications.
+ * * 'enabledNonBlocking' - (Default, for compatibility.) The initial navigation starts after the
+ * root component has been created. The bootstrap is not blocked until the initial navigation is
+ * complete.
+ * * 'enabled' - Same as `enabledBlocking`, will be removed in v11. @deprecated
```
Comment (atscott) [inline packages/router/src/router_module.ts:242]: You'll have to wait until v12 to remove: https://angular.io/guide/releases#deprecation-practices
Please also add an entry to `deprecations.md`

File: packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts (hunk 24 above)
Comment (atscott) [reply packages/core/schematics/migrations/initial-navigation/google3/tslint_update_recorder.ts:14]: I got the tests to pass by changing `false` to `enableNonBlocking`. This was more in line with the application's needs than `disabled`.

**Comment by atscott** [comment]
caretaker note: This will require a g3 patch - https://critique-ng.corp.google.com/cl/309442925

File: packages/router/src/router_module.ts (hunk 28 above)
Comment (CaerusKaru) [reply packages/router/src/router_module.ts:242]: Based on the deprecations file, it actually seems like v13. I've adjusted accordingly; I for one am in no rush to remove it.

File: packages/core/schematics/migrations/initial-navigation/transform.ts [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+  /** Migrate the ExtraOptions#InitialNavigation property assignments. */
+  migrateInitialNavigationAssignments(literals: ts.PropertyAssignment[]) {
+    literals.forEach(l => this.migrateAssignment(l));
+  }
+
+  /** Migrate an ExtraOptions#InitialNavigation expression to use the new options format. */
+  migrateAssignment(assignment: ts.PropertyAssignment) {
+    const newAssignment = ts.updatePropertyAssignment(
+        assignment, assignment.name, getUpdatedInitialNavigationValue(assignment.initializer));
+    this._updateNode(assignment, newAssignment);
```
Comment (devversion) [inline packages/core/schematics/migrations/initial-navigation/transform.ts:29]: One last thing: Could we update the node if the initial navigation value changed? Otherwise people will see unnecessary changes in their app, and potential formatting changes.

**Review by devversion — APPROVED** [review]
One minor last thing for schematics. Great work! LGTM on schematics.

**Comment by atscott** [comment]
[global presubmit](https://test.corp.google.com/ui#id=OCL:308924214:BASE:309422005:1588351430655:978697af) with patch https://critique-ng.corp.google.com/cl/309442925

File: packages/core/schematics/migrations.json [hunk 30]
```diff
@@ -44,6 +44,11 @@
       "version": "9.0.0-beta",
       "description": "ModuleWithProviders migration. In Angular 9, the ModuleWithProviders type without a generic has been deprecated. This migration adds the generic where it is missing. Read more about this here: https://v9.angular.io/guide/migration-module-with-providers",
       "factory": "./migrations/module-with-providers/index"
+    },
+    "migration-v10-router-initial-navigation-options": {
```
Comment (kara) [inline packages/core/schematics/migrations.json:48]: Can we move this migration into a follow-up PR? The PR is getting pretty meaty, and this migration requires an additional guide in the docs.

File: aio/content/guide/deprecations.md [hunk 31]
```diff
@@ -46,6 +46,7 @@ v9 - v12
 | `@angular/core`               | [Undecorated base classes that use Angular features](#undecorated-base-classes) | <!--v9--> v10 |
 | `@angular/forms`              | [`ngModel` with reactive forms](#ngmodel-reactive)                            | <!--v6--> v10 |
 | `@angular/router`             | [`preserveQueryParams`](#router)                                              | <!--v7--> v10 |
+| `@angular/router`             | [`InitialNavigation`](#router)                                                | <!--v10--> v13 |
```
Comment (kara) [inline aio/content/guide/deprecations.md:49]: IIUC, we're not removing `InitialNavigation` itself, just some of its options, so we should be more specific.

```suggestion
| `@angular/router`             | [`InitialNavigation` legacy options](#router)                                                | <!--v10--> v13 |
```

File: aio/content/guide/deprecations.md [hunk 32]
```diff
@@ -125,6 +126,7 @@ Tip: In the [API reference section](api) of this doc site, deprecated APIs are i
 | API | Replacement | Deprecation announced | Notes |
 | --- | ----------- | --------------------- | ----- |
 | [`preserveQueryParams`](api/router/NavigationExtras#preserveQueryParams) | [`queryParamsHandling`](api/router/NavigationExtras#queryParamsHandling) | v4 | none |
+| [`InitialNavigation`](api/router/InitialNavigation) | `enabled` option | v10 | The new option with 1:1 parity is `enabledBlocking`. |
```
Comment (kara) [inline aio/content/guide/deprecations.md:129]: This says that the replacement API is the `enabled` option. My understanding is that the `enabled` option is what is being deprecated. Fix?

The convention is also to have a section in this guide explaining the deprecation (why, what the replacement is, etc). Can you create the section and link to it in the "notes" column here?

**Review by kara — CHANGES_REQUESTED** [review]
Thanks!

**Comment by atscott** [comment]
Closing as outdated. I believe we made this change in a different PR.
